keps/sig-apps/0008-topology-aware-workload-controllers.md

```diff
@@ -0,0 +1,350 @@
+
+However, when scaling down, the controller decides which pods to kill first,
+without going through default scheduler. Therefore, scaling down ignores
+scheduling policy. 
+
+Controller rolling update is done by killing old replicas and creating new ones.
+Similar to scaling down cases, when old replicas are brought down, the decision
+doesn’t go through the default scheduler, either. On the other hand, the
+creation of new replicas goes through the default scheduler, so the spreading
+only works after the rolling update is done. 
+
+### Update Patterns 
+
+The workload controllers that support rolling update feature offer customizable
+update strategies. Users can specify the number of old replicas to bring down or
+the number of new replicas to create during a rollout. 
+
+Existing update patterns are not topology-aware. For example, users cannot
+specify things like: 
+
+* Update by failure domain: Update one failure domain at a time (staged domain)
+* Update across failure domains: Only allow X number of replicas to be down per
+  failure domain 
+
+## Proposal
+
+We can think of updates as planned disruptions. When a workload controller
+performs a rolling update, it will evict (instead of deleting) old pods and
+creates new pods. Users will be able to use [`PodDisruptionBudget`][](PDB) to
+control all sorts of planned disruptions, including updates. 
+
+To achieve this, PDB should allow users to specify allowed pod disruptions for
+a collection of nodes (such as the nodes in one failure domain), using a new PDB
+field `.spec.nodeSelectorTerm`. 
+
+This requires changes to workloads controllers code (change pod deletions to
+evictions), PDB API, and PDB controller code.
+
+[`PodDisruptionBudget`]: https://kubernetes.io/docs/concepts/workloads/pods/disruptions/
+
+### User Stories
+
+#### Story 1: Update By Failure Domain 
+
+When a user wants to update one failure domain at a time, they can create a PDB
+to disallow any disruptions in failure domains other than the first domain to
+update:
+
+```yaml
+apiVersion: policy/v1beta1
+kind: PodDisruptionBudget
+metadata:
+  name: app-pdb
+spec:
+  maxUnavailable: 0
+  selector:
+    matchLabels:
+      app: foo
+  nodeSelectorTerm:
+      # Only allow evicting/updating pods in zone “bar”
+      matchExpressions:
+      - key: failure-domain.beta.kubernetes.io/zone
+        operator: NotIn
+        values: 
+        - bar
+      - key: failure-domain.beta.kubernetes.io/zone
+        operator: Exists
+```
+
+#### Story 2: Update Across Failure Domain (PDB Per Failure Domain)
+
+When a user wants to update across failure domains, and only allow one pod per
+failure domain to be unavailable at a time, they can specify:
+
+```yaml
+apiVersion: policy/v1beta1
+kind: PodDisruptionBudget
+metadata:
+  name: app-pdb
+spec:
+  maxUnavailable: 1
+  selector:
+    matchLabels:
+      app: foo
+  nodeSelectorTerm:
+      # Only allow 1 pod to be evicted/updated per zone
```

Comment:
How does this allow 1 pod to be evicted per zone? The matchExpression wants the node to have the "zone" label and all the nodes of the cluster have such label. So, I think this rule causes only 1 pod to be evicted in the whole cluster, not 1 pod per zone.